src/shims/unix/fd.rs

```diff
@@ -273,6 +273,35 @@ impl FdTable {
 
 impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
 pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    fn dup(&mut self, old_fd: i32) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        match this.machine.fds.dup(old_fd) {
+            Some(dup_fd) => Ok(this.machine.fds.insert_fd_with_min_fd(dup_fd, 0)),
+            None => this.fd_not_found(),
+        }
+    }
+
+    fn dup2(&mut self, old_fd: i32, new_fd: i32) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        match this.machine.fds.dup(old_fd) {
+            Some(dup_fd) => {
+                if new_fd != old_fd {
+                    // Close new_fs if it is previously opened.
+                    if let Some(file_descriptor) = this.machine.fds.remove(new_fd) {
+                        // Ignore close error according to dup2() doc.
+                        file_descriptor.close(this.machine.communicate()).ok();
+                    }
+
+                    this.machine.fds.fds.insert(new_fd, dup_fd);
```

Comment:
`insert` returns the old value if any, so instead of calling `remove` first you can just use that.